### PR TITLE
Thread processor affinity

### PR DESCRIPTION
**Pull request by mikbras**
Add support for the following syscalls:

- **SYS_sched_getaffinity** - delegated to target
- **SYS_sched_setaffinity** - delegated to target
- **SYS_getcpu** - delegated to target
- **SYS_mbind** - stub only for now (to enable Python webserver solution)

Also adds a way to map from a **kernel thread TID** to a **target thread TID**.

### CODE REVIEW

File: tests/ltp/Makefile [hunk 1]
```diff
@@ -185,6 +185,3 @@ one:
 	$(MYST) mkcpio ramfs.appdir ramfs
 	sudo -E $(RUNTEST) $(MYST_EXEC) $(OPTS) $(FS) $(TEST) $(NL)
 endif
-
```
Comment (bodzhang) [inline tests/ltp/Makefile:188]: Is the deletion here intentional?

File: include/myst/thread.h [hunk 2]
```diff
@@ -88,6 +88,9 @@ struct myst_thread
     /* unique thread identifier (same as pid for main thread) */
     pid_t tid;
 
+    /* the value returned by gettid() on the target (for this thread) */
```
Comment (bodzhang) [inline include/myst/thread.h:91]: `myst_gettid()` still returns the internal `tid` above. `myst_sched_getaffinity()` and myst_sched_setaffinity()' perform the internal-tid to `target_tid` translation. The comment should be fixed.

**Review by bodzhang — CHANGES_REQUESTED** [review]

File: solutions/aspnet/Makefile [hunk 3]
```diff
@@ -4,11 +4,11 @@ include $(TOP)/defs.mak
...
 
-#ifdef STRACE
-#OPTS += --strace
-#endif
+ifdef STRACE
+OPTS += --strace
+endif
 
-OPTS = --memory-size=256m
+OPTS += --memory-size=2048m
```
Comment (jxyang) [inline solutions/aspnet/Makefile:11]: This setting should has no effect as the same setting in config.json takes precedence.

File: solutions/aspnet/Makefile [hunk 4]
```diff
@@ -17,7 +17,7 @@ run: appdir private.pem
 	test -f server.pid && kill -9 `cat server.pid` || true
 	$(MYST) package appdir private.pem config.json
 	myst/bin/$(APPNAME) $(OPTS) & echo $$! > server.pid
-	sleep 15
+	sleep 60
```
Comment (jxyang) [inline solutions/aspnet/Makefile:20]: 60 seconds is probably too much for waiting. Why it takes much longer now to boot Mystikos?

File: solutions/python_webserver/Makefile [hunk 5]
```diff
@@ -7,7 +7,7 @@ ifdef STRACE
 OPTS = --strace
 endif
 
-OPTS = --memory-size=512m
+OPTS = --memory-size=1024m
```
Comment (jxyang) [inline solutions/python_webserver/Makefile:10]: See my comment above.

File: include/myst/kernel.h [hunk 6]
```diff
@@ -95,6 +95,9 @@ typedef struct myst_kernel_args
     /* The number of threads that can be created (including the main thread) */
     size_t max_threads;
 
+    /* The tid/pid of the main thread passed from the host */
+    pid_t target_tid;
```
Comment (jxyang) [inline include/myst/kernel.h:99]: This applies to all ethreads. So the comment about just "main thread" is not accurate.

File: tools/myst/enc/enc.c [hunk 7]
```diff
@@ -735,7 +740,7 @@ int myst_load_fssig(const char* path, myst_fssig_t* fssig)
 #define ENCLAVE_PRODUCT_ID 1
 #define ENCLAVE_SECURITY_VERSION 1
 #define ENCLAVE_DEBUG true
-#define ENCLAVE_HEAP_SIZE 131072
+#define ENCLAVE_HEAP_SIZE 262144
```
Comment (jxyang) [inline tools/myst/enc/enc.c:743]: Is this the OE heap? Why does it need to double?

**Review by bodzhang — APPROVED** [review]

**Comment by mikbras** [comment]
I found that it takes 5 minutes on a bare metal SGX NUC for the ``aspnet`` solution to run. Previously it took only 1 minute. To work around this, I added the UNIPROCESSOR macro. When defined ``sched_getaffinity`` only reports a single processor, causing ``aspnet`` to run in 60 seconds again.

File: include/myst/mman.h [hunk 8]
```diff
@@ -140,4 +150,13 @@ int myst_mman_free_size(myst_mman_t* mman, size_t* size);
 
 void myst_mman_dump_vads(myst_mman_t* mman);
 
+int myst_mman_mprotect(myst_mman_t* mman, void* addr, size_t len, int prot);
```
Comment (jxyang) [inline include/myst/mman.h:153]: Do we have to mingle `mprotect` work in this PR?

File: samples/openmp/Makefile [hunk 9]
```diff
@@ -19,10 +19,10 @@ rootfs: appdir
 	# $(MYST) mkcpio appdir rootfs
 
 run:
-	$(MYST_EXEC) $(OPTS) --roothash=roothash ext2rootfs /app/openmp-test
+	OMP_PROC_BIND=true $(MYST_EXEC) $(OPTS) --roothash=roothash ext2rootfs /app/openmp-test
```
Comment (jxyang) [inline samples/openmp/Makefile:22]: What's this for?

**Pull request closed without merging**